### PR TITLE
Draft PR - CSV Harvester with check schema validation and Geonetwork update

### PR DESCRIPTION
**Pull request by ndhogan**
Draft PR so @mhidas can comment on code changes, bugs identified, etc.

----
Confirmed: python-testcontainers (postgresql) is compatible with github actions containers

### CODE REVIEW

**Comment by ndhogan** [comment]
@mhidas - in order to check the spatial extent update you may need to point the _Geonetwork3.artifact-source_ in your aodnstack to _s3prefix://imos-artifacts/builds/geonetwork-build_build/master/53/artifacts/main/target/geonetwork-imos-1.0.45.war_, eg.

```yaml
  Geonetwork3:
      CloudformationTemplateURL: ebapps/geonetwork3-template.yaml
      application: geonetwork3
      environment: dev
      emptyrootcontext: true
      # update this property
      artifact-source: s3prefix://imos-artifacts/builds/geonetwork-build_build/master/53/artifacts/main/target/geonetwork-imos-1.0.45.war
      ...
```

**Comment by mhidas** [comment]
:+1: Awesome! Thanks @ndhogan . I'll have a play with this again tomorrow.

**Comment by ndhogan** [comment]
Format for _harvest_params.metadata_updates_:

```json
"metadata_updates": [
    {
        "uuid": "sample-acorn",
        "spatial": {
            "table": "string",
            "column": "string",
            "resolution": 0
        },
        "temporal": {
            "table": "string",
            "column": "string"
        },
        "vertical": {
            "table": "string",
            "column": "string"
        }
    }
]
```

_Note: __column__ attributes are case sensitive - as these are passed into the query strings within double quotes._

Full example below:

```json

{
    "id": "BGC",
    "path": [ "BGC" ],
    "handler": "BgcHandler",
    "params": {
        "include_regexes": [
            ".*\\.csv$"
        ],
        "harvest_params": {
            "db_schema": "bgc",
            "db_objects": [
                {
                    "name": "nrs_bgc_chemistry_data",
                    "type": "table"
                },
                {
                    "name": "bgc_phyto_changelog",
                    "type": "table"
                },
                {
                    "name": "bgc_phyto_raw",
                    "type": "table"
                },
                {
                    "name": "bgc_trip",
                    "type": "table"
                },
                {
                    "name": "bgc_zoop_changelog",
                    "type": "table"
                },
                {
                    "name": "bgc_zoop_raw",
                    "type": "table"
                },
                {
                    "name": "bgc_phyto_aggregation",
                    "type": "materialized view",
                    "dependencies": ["bgc_phyto_raw"]
                },
                {
                    "name": "bgc_zoop_aggregation",
                    "type": "materialized view",
                    "dependencies": ["bgc_zoop_raw"]
                }
            ],
            "metadata_updates": [
                {
                    "uuid": "sample-acorn",
                    "spatial": {
                        "table": "nrs_bgc_chemistry_data",
                        "column": "geom",
                        "resolution": 3
                    },
                    "temporal": {
                        "table": "nrs_bgc_chemistry_data",
                        "column": "LOCAL_TRIP_START_TIME"
                    },
                    "vertical": {
                        "table": "nrs_bgc_chemistry_data",
                        "column": "SAMPLE_DEPTH_M"
                    }

                }
            ]
        }
    }
}
```

**Comment by ndhogan** [comment]
Still need to add unit tests for some of the new functions, and add SSL to the database connection method.

**Comment by mhidas** [comment]
I tried to run up a stack with the Geonetwork artefact updated as [above](https://github.com/aodn/python-aodncore/pull/237#issuecomment-891405458). It set up most of the stack successfully, but got an error about geonetwork:
```
TASK [geonetwork3_config : update settings] ********************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: 
ansible.module_utils.geonetwork3lib.GeonetworkHTTPError: GeonetworkHTTPError(message='HTTPError('400 Client Error:  for url: http://geonetwork3-aodnstack-mhidas.dev.aodn.org.au/geonetwork/srv/api/0.1/site/settings',)', status_code='400', error_class='None', inner_exception='400 Client Error:  for url: http://geonetwork3-aodnstack-mhidas.dev.aodn.org.au/geonetwork/srv/api/0.1/site/settings')
```

**Review by mhidas — COMMENTED** [review]
In general it's working well - just a few comments and suggestions so far.

I haven't tested the metadata update functionality yet.

File: aodncore/pipeline/db.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        if fn and step['type'] == 'table':
+            self._logger.info("Creating {type} {name}".format(**step))
+            try:
+                with open(fn) as stream:
+                    try:
+                        schema = get_tableschema_descriptor(yaml.safe_load(stream), 'schema')
+                        columns = []
+                        for f in schema['fields']:
+                            f['type'] = 'geometry(Geometry,4326)' if f['name'] == 'geom' else get_field_type(f['type'])
+                            columns.append('"{name}" {type}'.format(**f))
```
[the remark this answers is not in the document]
Comment (ndhogan) [reply aodncore/pipeline/db.py]: I did kind of assume that if it was upper- or mixed-case in the yaml file then we would want that to be reflected in the database, but if that's not the case we can remove the double-quotes?

File: aodncore/pipeline/db.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+                self.__exec(stream.read())
+
+    def create_table_from_yaml_file(self, step):
+        """Function to read an yaml file and use it to build a CREATE TABLE script for execution against the database.
+
+        :param step: A dict containing 'name' and 'type' (at least) keys
+        - step.name is the name used as part of the match regular expression
+        - step.type is the type of database object. Type should always be table in this context
+        """
+        fn = find_file(self.schema_base_path, '(.*){}(.*).yaml'.format(step['name']))
```
[the remark this answers is not in the document]
Comment (ndhogan) [reply aodncore/pipeline/db.py:185]: sounds like a plan - also I guess we should also accommodate both _yaml_ and _yml_ ?

File: aodncore/pipeline/db.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+    def __exec_copy(self, statement, file):
+        """Execute a COPY FROM statement using the instance cursor.
+
+        :param statement: A string containing a COPY statement.
+        :param file: A readable file-like object.
+        :return: None
+        """
+        try:
+            self._cur.copy_expert(sql.SQL(statement), file)
```
[the remark this answers is not in the document]
Comment (ndhogan) [reply aodncore/pipeline/db.py:96]: this issue apppears to be mentioned here: https://github.com/psycopg/psycopg2/issues/1141#issuecomment-687661063 - looks like postgresql doesn't return PQcmdStatus for copy operations - i'll just add query

File: aodncore/pipeline/db.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhidas) [reply aodncore/pipeline/db.py]: Yep, good point

**Comment by mhidas** [comment]
A perfect example of why this is the preferred behaviour is https://github.com/aodn/issues/issues/1060, where all the pipelines are failing just because they can't connect to Geonetwork

File: aodncore/pipeline/steps/harvest.py [hunk 4]
```diff
@@ -403,5 +411,125 @@ def run_additions(self, harvester_map, tmp_base_dir):
...
+                password=gn_config['metadata_password']
+            )
+            with DatabaseInteractions(config=self.get_config_file('database.json'),
+                                      schema_base_path=self._config.pipeline_config['harvester']['schema_base_dir'],
+                                      logger=self.logger) as conn:
+                for m in metadata:
+                    if 'spatial' in m:
+                        m['spatial']['db_schema'] = self.params['db_schema']
+                    handler = GeonetworkMetadataHandler(conn, gn, m, self.logger)
+                    handler.run()
```
Comment (ndhogan) [inline aodncore/pipeline/steps/harvest.py:471]: @mhidas re:

> Only one issue with it: if this part fails, it currently raises an unhandled exception, so the pipeline ends up in an ERROR state. This is despite having successfully harvested the data. I think the desired behaviour in such a case would be to only issue a warning, but allow the pipeline to end with a SUCCESS status.

There are a number of potential exceptions in the classes called in this section of _CsvHarvesterRunner.run()_, so we could probably move it to the end of the method and embed it in a try statement - catch any exception and just log it a warning

Comment (mhidas) [reply aodncore/pipeline/steps/harvest.py:471]: Yep, that would be good. In fact we could be a bit picky about which exceptions we catch. In particular, I think anything to do with talking to Geonetwork could be caught.
Exceptions arising from querying the database could be allowed through, as they might indicate a config error or something more sinister.

Comment (ndhogan) [reply aodncore/pipeline/steps/harvest.py:471]: so more like...

```python
            except [InvalidSQLConnectionError, InvalidSQLTransactionError, MissingFileError]:
                raise
            except Exception as e:
                self._logger.warning(e)
```

Comment (ndhogan) [reply aodncore/pipeline/steps/harvest.py:471]: I've created [fix-geonetwork-error-fail](https://github.com/aodn/python-aodncore/tree/fix-geonetwork-error-fail) as a fork of [csv-harvester](https://github.com/aodn/python-aodncore/tree/csv-harvester) and committed the above change to that branch - but I won't be able to test it before I go on leave...

File: aodncore/pipeline/steps/harvest.py [hunk 5]
```diff
@@ -403,5 +411,125 @@ def run_additions(self, harvester_map, tmp_base_dir):
...
+            runsheet = (x for x in map(self.build_runsheet, objs) if x)
+        else:
+            raise MissingConfigParameterError('Generic CSV Harvester requires that the '
+                                              'harvest_params["db_objects"] attribute be set')
+
+        with DatabaseInteractions(config=self.get_config_file('database.json'),
+                                  schema_base_path=self._config.pipeline_config['harvester']['schema_base_dir'],
+                                  logger=self.logger) as conn:
+            process = self.get_process_sequence(conn)
+            for step in runsheet:
```
[the remark this answers is not in the document]
Comment (mhidas) [reply aodncore/pipeline/steps/harvest.py:447]: Fixed in #242

**Comment by mhidas** [comment]
All the review feedback here has been addressed and merged into the wsd branch and we'll merge to master from there (#239).
I'll close this PR to avoid confusion.

**Pull request closed without merging**